java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java

```diff
@@ -112,9 +114,28 @@ private TypeSubstitution inferTypeSubstitutionInParameterizedType(MethodJavaSymb
     List<JavaType> formalTypeSubstitutedTypes = formalType.typeSubstitution.substitutedTypes();
     TypeSubstitution result = substitution;
     if (argType.isParameterized()) {
-      List<JavaType> argTypeSubstitutedTypes = ((ParametrizedTypeJavaType) argType).typeSubstitution.substitutedTypes();
-      TypeSubstitution newSubstitution = inferTypeSubstitution(method, formalTypeSubstitutedTypes, argTypeSubstitutedTypes);
-      result = mergeTypeSubstitutions(substitution, newSubstitution);
+      ParametrizedTypeJavaType parametrizedArgType = (ParametrizedTypeJavaType) argType;
+      if(parametrizedArgType.rawType == formalType.rawType) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)